comms/core/src/net_address/multiaddr_with_stats.rs

```diff
@@ -183,6 +183,7 @@ impl MultiaddrWithStats {
     /// Reset the connection attempts on this net address for a later session of retries
     pub fn reset_connection_attempts(&mut self) {
         self.connection_attempts = 0;
+        self.last_failed_reason = None;
```

Comment:
This works hand in hand when trying to connect in `async fn dial_peer_with_retry`, to start with a clean slate. If there is any error on one of the multi addresses, it will be populated again. We can then use `if peer.all_addresses_failed() {` instead of compound logic `if peer.last_connect_attempt().is_some() && peer.last_seen().is_none() {` to determine if we need to try again.

Reply:
In `Peer`
```rust
    /// Provides info about the failure status of all addresses
    pub fn all_addresses_failed(&self) -> bool {
        self.addresses.iter().all(|a| a.last_failed_reason().is_some())
    }
```